components/raftstore/src/store/read_queue.rs

```diff
@@ -230,14 +216,47 @@ where
         None
     }
 
-    pub fn advance_leader_reads<T>(&mut self, states: T)
+    pub fn advance_leader_reads<T>(&mut self, tag: &str, states: T)
     where
         T: IntoIterator<Item = (Uuid, Option<LockInfo>, u64)>,
     {
-        for (uuid, _, index) in states {
-            assert_eq!(uuid, self.reads[self.ready_cnt].id);
-            self.reads[self.ready_cnt].read_index = Some(index);
-            self.ready_cnt += 1;
+        let mut states_iter = states.into_iter();
+        while let Some((uuid, info, index)) = states_iter.next() {
+            let invalid_id = match self.reads.get_mut(self.ready_cnt) {
+                Some(r) if r.id == uuid => {
+                    r.read_index = Some(index);
+                    self.ready_cnt += 1;
+                    continue;
+                }
+                Some(r) => Some((r.id, r.propose_time)),
+                None => None,
+            };
+
+            if !self.contexts.contains_key(&uuid) {
```

Comment:
Only non-leader requests will be put into `self.context`, so leader can always skip panic due to this check :
https://github.com/tikv/tikv/blob/356774c4a9e6d75f6cd589694b982130e017428b/components/raftstore/src/store/read_queue.rs#L209-L213

Reply:
I deleted the block. As there is a message filter before stepping, the block will not be triggered for any known cases.